lib/OpenQA/Task/Needle/Save.pm

```diff
@@ -0,0 +1,179 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::Task::Needle::Save;
+use Mojo::Base 'Mojolicious::Plugin';
+
+use OpenQA::Jobs::Constants;
+use OpenQA::Utils;
+use Mojo::JSON 'decode_json';
+
+sub register {
+    my ($self, $app) = @_;
+    $app->minion->add_task(save_needle => sub { _save_needle($app, @_) });
+}
+
+sub _json_validation {
+    my ($json) = @_;
+
+    my $djson = eval { decode_json($json) };
+    if (!$djson) {
+        my $err = $@;
+        $err =~ s@at /usr/.*$@@;    # do not print perl module reference
+        die "syntax error: $err";
+    }
+    if (!exists $djson->{area} || !exists $djson->{area}[0]) {
+        die 'no area defined';
+    }
+    if (!exists $djson->{tags} || !exists $djson->{tags}[0]) {
+        die 'no tag defined';
+    }
+
+    my $areas = $djson->{area};
+    foreach my $area (@$areas) {
+        die 'area without xpos'   unless exists $area->{xpos};
+        die 'area without ypos'   unless exists $area->{ypos};
+        die 'area without type'   unless exists $area->{type};
+        die 'area without height' unless exists $area->{height};
+        die 'area without width'  unless exists $area->{width};
+    }
+    return $djson;
+}
```

Comment:
This would seem like a nice use case for `JSON::Validator` actually.

Reply:
Yes, but I'd save this for later.